### PR TITLE
SONiC CLI Auto-generation tool HLD

### PR DESCRIPTION
**Pull request by vadymhlushko-mlnx**
| PR title | state | context |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| [[sonic-cli-gen] first phase implementation of the SONiC CLI Auto-generation tool](https://github.com/Azure/sonic-utilities/pull/1644) | GitHub issue/pull request detail | GitHub pull request check contexts |
| [[sonic-package-manager] support sonic-cli-gen and packages with YANG model](https://github.com/Azure/sonic-utilities/pull/1650) | GitHub issue/pull request detail | GitHub pull request check contexts |

Signed-off-by: Vadym Hlushko <vadymh@nvidia.com>
The high-level design document for the SONiC CLI Auto-generation tool.

### CODE REVIEW

**Comment by anshuv-mfst** [comment]
@prsunny @lguohan - could you please take a look, thanks.

**Comment by vadymhlushko-mlnx** [comment]
@praveen-li thanks for your comments, but I can't see to which line in the HLD your comments are related?

**Comment by vadymhlushko-mlnx** [comment]
> https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-interface.yang
> 
> How does the auto-generation of CLI config/show commands would look like for the above yang when we have a table container with two lists? please provide some examples for clarity.

Could you please take a look at the `Open questions` section and give your feedback?

**Comment by vadymhlushko-mlnx** [comment]
> If possible we can put one real example.

I made real examples for some auto-generation rules, could you please check?

**Comment by liat-grozovik** [comment]
@venkatmahalingam and @praveen-li thanks for the comments.
Could you please review the update HLD? we would like to upstream the code aligned with it in the next 1-2 weeks so HLD should be closed before that.

**Comment by liat-grozovik** [comment]
@venkatmahalingam and @praveen-li as the feature is now ready for review appreciate if you can check comments handling and approve the HLD as well.

**Comment by vadymhlushko-mlnx** [comment]
@praveen-li could you please also take a look at the implementation PR's (you can find them in this PR description)?

**Comment by praveen-li** [comment]
Hi Team

I have to take care an urgent LinkedIn Internal Task. I will not be able to review it until 06/15/20221. Sorry about it.


Thanks,

Regards
Praveen
________________________________

**Comment by liat-grozovik** [comment]
@praveen-li appreciate your feedback

File: doc/cli_auto_generation/cli_auto_generation.md [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+`vlanid`, `mtu`, `admin-status` - are not-positional arguments, and to give them the next style MUST be used (check `config command`)
+__This style "--arg" is NOT RELATED to the Linux CLI optional parameter style__
+
+###### `config` command
+```
+admin@sonic:~$ config vlan add Vlan11 --vlanid 11 --mtu 128 --admin-status up
+admin@sonic:~$ config vlan del Vlan11
+```
+
+The `update` command will be generated but not always could be executed. The user will get an error message from the YANG validator if the `update command` is not supported.
```
Comment (jleveque) [inline doc/cli_auto_generation/cli_auto_generation.md:344]: There are a few references to the `update` subcommand, but no examples. Would it be possible to provide an example of the `update` subcommand, if it is supported?

Comment (vadymhlushko-mlnx) [reply doc/cli_auto_generation/cli_auto_generation.md:344]: fixed

**Review by jleveque — APPROVED** [review]

**Comment by rlhui** [comment]
@vadymhlushko-mlnx, would you please confirm the code PRs for this feature are the two PRs listed above?

**Comment by liat-grozovik** [comment]
> @vadymhlushko-mlnx, would you please confirm the code PRs for this feature are the two PRs listed above?

Yes it is.

**Pull request merged**